File: iitj_assignments/assignment_1/c programs/5_stack.c
#include<stdio.h>
#include<stdlib.h>
#include <stdbool.h>
//stack node
struct stack
{
  int top;
  int capacity;
  int *array;
};

//Initialization
struct stack* creation()
{
  struct stack *s;
  s->capacity =100;
  s->top= -1;
  s->array = (int *)malloc(s->capacity*sizeof(int));
  return s;
}
//stack is full or not
bool isFull(struct stack *s)
{
  if(s->top==s->capacity-1)
    return true;
  return false;
}
//stack is empty or not
bool isEmpty(struct stack *s)
{
  if(s->top==-1)
    return true;
  return false;
}
//Pust element
void push(struct stack *s,int data)
{
  if(isFull(s))
    printf("Overflow\n");
  else
      s->array[++s->top]=data;

}
//pop element
int pop(struct stack *s)
{
  if(isEmpty(s))
    {
      printf("underflow\n");
      return -1;
    }
  else
     return s->array[s->top--];

}
//get top element
int getTop(struct stack *s)
{
  if(isEmpty(s))
    {
      printf("underflow\n");
      return -1;
    }
  else
  return s->array[s->top];
}
//display stack
void display(struct stack *s)
{
  if(isEmpty(s))
      printf("underflow\n");
  else
  {
    for(int i=0;i<=s->top;i++)
       printf("%d\t",s->array[i]);
    printf("\n");
  }
}
//main function
int main()
{
    int choice,data,temp;
    struct stack *s=NULL;
    s=creation();
    while(1)
    {
        printf("*************************************\n");
        printf("1.push data\n");
        printf("2.pop data\n");
        printf("3.display\n");
        printf("4.get top\n");
        printf("5.Exit\n");
        printf("enter your choice \n");
        scanf("%d",&choice );
        switch(choice)
        {
            case 1:
                    printf("enter your data \n");
                    scanf("%d",&data );
                    push(s,data);
                    break;
            case 2:
                    temp = pop(s);
                    if(temp!=-1)
                       printf("popped element is %d\n",temp);
                    break;
            case 3:
                   display(s);
                   break;
            case 4: printf("top element is  %d\n",getTop(s));
                    break;

            case 5:
                    exit(0);
            default:
                    printf("wrong choice\n");
                    break;
        }
    }
}

File: iitj_assignments/assignment_1/python programs/5_stack.py
#!/usr/bin/env python
# coding: utf-8

# In[14]:


class node:
    def __init__(self,data):
        self.data = data
        self.next = None
class Stack:
    #initiation
    def __init__(self):
        self.head = None
    #push element at top
    def push(self,data):  #inserting at start of linked list
        newnode=node(data)
        if self.head == None:
            self.head=newnode

        else:
            newnode.next = self.head
            self.head = newnode
    #pop element from top
    def pop(self):  #delete from head of linked list
        if self.isEmpty():
            return None

        else:
            temp = self.head.data
            self.head = self.head.next
            return temp

    # Returns the head node data
    def peek(self):
        if self.isEmpty():
            return None
        else:
            return self.head.data
    #display stack
    def display(self):

        temp = self.head
        if self.isEmpty():
            print("Stack Underflow")

        else:
            while(temp != None):
                print(temp.data,end = " ")
                temp = temp.next
    #stack is empty or not        
    def isEmpty(self):
        if self.head == None:
            return True
        else:
            return False
s=Stack()
s.push(1)
s.push(2)
s.push(3)
s.push(4)

# Display stack elements
s.display()
print(' ')

print("Top element is ",s.peek())


print("Popped element is ", s.pop())

print("Popped element is ", s.pop())

s.display()
